### ISSUE TITLE
Panic from logrus while serving the request

### ISSUE THREAD

**Issue by ahmetalpbalkan** [docker/distribution#2024]
I saw this in the logs. Any ideas what it might be?

```
2016/10/28 20:30:34 http: panic serving 172.17.0.1:52442: &{0xc420044ac0 map[] 2016-10-28 20:30:34.103423382 +0000 UTC panic runtime error: invalid memory address or nil pointer dereference}
goroutine 13201 [running]:
net/http.(*conn).serve.func1(0xc420561700)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:1491 +0x12a
panic(0xb6bc00, 0xc42032cdc0)
    /usr/local/Cellar/go/1.7.1/libexec/src/runtime/panic.go:458 +0x243
github.com/docker/distribution/vendor/github.com/Sirupsen/logrus.(*Entry).log(0xc42032cdc0, 0x0, 0xc42032ce40, 0x40)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/Sirupsen/logrus/entry.go:103 +0x38d
github.com/docker/distribution/vendor/github.com/Sirupsen/logrus.(*Entry).Panic(0xc42032cdc0, 0xc420c38a60, 0x1, 0x1)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/Sirupsen/logrus/entry.go:148 +0xed
github.com/docker/distribution/vendor/github.com/Sirupsen/logrus.(*Logger).Panic(0xc420044ac0, 0xc420c38a60, 0x1, 0x1)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/Sirupsen/logrus/logger.go:155 +0xf4
github.com/docker/distribution/vendor/github.com/Sirupsen/logrus.Panic(0xc420c38a60, 0x1, 0x1)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/Sirupsen/logrus/exported.go:102 +0x4b
github.com/docker/distribution/registry.panicHandler.func1.1()
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/registry.go:297 +0x116
panic(0xab66e0, 0xc420012030)
    /usr/local/Cellar/go/1.7.1/libexec/src/runtime/panic.go:458 +0x243
github.com/docker/distribution/vendor/github.com/Azure/azure-sdk-for-go/storage.BlobStorageClient.DeleteBlobIfExists(0xc420552480, 0x18, 0xc421350190, 0x40, 0x42, 0x1, 0xc420357af0, 0x10, 0xb8426d, 0xa, ...)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/Azure/azure-sdk-for-go/storage/blob.go:929 +0x11c
github.com/docker/distribution/registry/storage/driver/azure.(*driver).PutContent(0xc4203dd190, 0xf108c0, 0xc42000c540, 0xc42000c4d0, 0x64, 0xc420552ae0, 0x14, 0x20, 0x0, 0x0)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/storage/driver/azure/azure.go:120 +0xd5
github.com/docker/distribution/registry/storage/driver/base.(*Base).PutContent(0xc4203dd1a0, 0xf108c0, 0xc42000c540, 0xc42000c4d0, 0x64, 0xc420552ae0, 0x14, 0x20, 0x0, 0x0)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/storage/driver/base/base.go:102 +0x26e
github.com/docker/distribution/registry/storage.(*linkedBlobStore).Create(0xc4200f4080, 0xf10b40, 0xc42000c230, 0x0, 0x0, 0x0, 0xbf5616, 0xc41ffd56c2, 0x0, 0xf45320)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/storage/linkedblobstore.go:169 +0x556
github.com/docker/distribution/notifications.(*blobServiceListener).Create(0xc420552780, 0xf10b40, 0xc42000c230, 0x0, 0x0, 0x0, 0x0, 0x0, 0x40cc68, 0xc420c39210)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/notifications/listener.go:167 +0xaa
github.com/docker/distribution/registry/handlers.(*blobUploadHandler).StartBlobUpload(0xc42000c230, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/handlers/blobupload.go:121 +0x14a
github.com/docker/distribution/registry/handlers.(*blobUploadHandler).StartBlobUpload-fm(0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/handlers/blobupload.go:32 +0x48
net/http.HandlerFunc.ServeHTTP(0xc4203dcfc0, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:1726 +0x44
github.com/docker/distribution/vendor/github.com/gorilla/handlers.MethodHandler.ServeHTTP(0xc420180960, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/gorilla/handlers/handlers.go:35 +0x379
github.com/docker/distribution/registry/middleware/registry/acrcorrelation.headerWriterHandler.ServeHTTP(0xf0aa40, 0xc420180960, 0xf10a80, 0xc4213511d0, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/middleware/registry/acrcorrelation/middleware.go:48 +0x173
github.com/docker/distribution/registry/middleware/registry/acrcorrelation.(*headerWriterHandler).ServeHTTP(0xc420552560, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    <autogenerated>:1 +0x8b
github.com/docker/distribution/registry/handlers.(*App).dispatcher.func1(0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/handlers/app.go:697 +0x741
net/http.HandlerFunc.ServeHTTP(0xc42037e1e0, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:1726 +0x44
github.com/docker/distribution/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc420298be0, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/gorilla/mux/mux.go:98 +0x255
github.com/docker/distribution/registry/handlers.(*App).ServeHTTP(0xc42036dc20, 0x7f22658c7448, 0xc421351180, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/handlers/app.go:615 +0x1e4
github.com/docker/distribution/registry.alive.func1(0xf0f8c0, 0xc4200d84e0, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/registry.go:317 +0x74
net/http.HandlerFunc.ServeHTTP(0xc420334810, 0xf0f8c0, 0xc4200d84e0, 0xc420132e10)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:1726 +0x44
github.com/docker/distribution/health.Handler.func1(0xf0f8c0, 0xc4200d84e0, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/health/health.go:271 +0x182
net/http.HandlerFunc.ServeHTTP(0xc42037e2c0, 0xf0f8c0, 0xc4200d84e0, 0xc420132e10)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:1726 +0x44
github.com/docker/distribution/registry.panicHandler.func1(0xf0f8c0, 0xc4200d84e0, 0xc420132e10)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/registry/registry.go:300 +0x81
net/http.HandlerFunc.ServeHTTP(0xc42037e2e0, 0xf0f8c0, 0xc4200d84e0, 0xc420132e10)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:1726 +0x44
net/http.serverHandler.ServeHTTP(0xc420264000, 0xf0f8c0, 0xc4200d84e0, 0xc420132e10)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:2202 +0x7d
net/http.(*conn).serve(0xc420561700, 0xf10540, 0xc4202b3a00)
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:1579 +0x4b7
created by net/http.(*Server).Serve
    /usr/local/Cellar/go/1.7.1/libexec/src/net/http/server.go:2293 +0x44d
```

It doesn't crash the process. I'm on `v2.3.0-rc.1` tree.

**Comment by ahmetalpbalkan** [comment]
I can actually reproduce this somewhat consistently by pounding the server with like 4 concurrent clients pushing all-new single-layer (5 mb blobsfrom urandom) images. It must have some impact on API QoS since the handling the request throws panic.

**Comment by dmcgowan** [comment]
@ahmetalpbalkan did you track this down? Based on stacktrace I assume the panic has nothing to do with logrus and is happening here ...
```
github.com/docker/distribution/vendor/github.com/Azure/azure-sdk-for-go/storage.BlobStorageClient.DeleteBlobIfExists(0xc420552480, 0x18, 0xc421350190, 0x40, 0x42, 0x1, 0xc420357af0, 0x10, 0xb8426d, 0xa, ...)
    /Users/alp/workspace/gopath-distribution/src/github.com/docker/distribution/vendor/github.com/Azure/azure-sdk-for-go/storage/blob.go:929 +0x11c
```

That file does not exist on `v2.3.0-rc.1` but looking at the latest version the panic makes sense
```
func (b BlobStorageClient) DeleteBlobIfExists(container, name string) (bool, error) {
	resp, err := b.deleteBlob(container, name)
	if resp != nil && (resp.statusCode == http.StatusAccepted || resp.statusCode == http.StatusNotFound) {
		return resp.statusCode == http.StatusAccepted, nil
	}
	defer resp.body.Close()   <----- resp could be nil
	return false, err
}
```

**Comment by ahmetalpbalkan** [comment]
My mistake. I got misleaded by the ordering of the goroutines. This is fixed now by #2027 actually.
